Fix: Adding collection form model validation errors to parent form

Dhamsoft opened the pull request.

Commit:
Add test coverage for validation fails on the child collection records

Commit:
Fix add collection forms validation to parent form

willbryant commented on lib/on_form/errors.rb:
```diff
@@ -4,7 +4,7 @@ def errors
     end
 
-  private
+    private
```
:(

willbryant commented on lib/on_form/form.rb:
```diff
@@ -81,7 +81,6 @@ def self.expose_collection_of(association_name, on: nil, prefix: nil, suffix: ni
       association_name = association_name.to_sym
 
       on = prepare_model_to_expose!(on)
-
```
?

willbryant commented on lib/on_form/form.rb:
```diff
@@ -91,8 +90,21 @@ def self.expose_collection_of(association_name, on: nil, prefix: nil, suffix: ni
+
       define_method("#{exposed_name}_attributes=") { |params| send(exposed_name).parse_collection_attributes(params) }
+      define_method("_validate_#{exposed_name}_forms") { send(exposed_name).validate_forms }
+      after_validation :"_validate_#{exposed_name}_forms"
```
Just wondering, since we have this callback, do you think we could/should add to the errors in that callback, instead of having to make the above change to add_collection_form_errors, which presumably meant you had to change the collection_wrappers stuff?

Dhamsoft replied:
Yep, That's a good idea. Unfortunately, we can't use after_validation. Instead of that, we have to use validate.

willbryant commented on test/collection_form_test.rb:
```diff
@@ -112,14 +112,31 @@ def initialize(house)
     room.area.must_equal 18
   end
 
+  it "returns false from valid? if a validation fails on the child records" do
```
Could we maybe add a test for validations that are defined on the form (as opposed to the backing records) failing?

Dhamsoft replied:
Added new test it block `returns false from valid? if a validation fails on the child form validation`"

Commit:
Adding collection form errors logic moveto CollectionWrapper class

- Update OnForm::Form.expose_collection_of to validate collection forms
  parent form object passed to CollectionWrapper to add errors back to parent form

- Update CollectionWrapper#validate_forms to receive parent_form as argument
  and add logic to add errors to parent if any

- Add test to check form validation errors added back to parent form

binary-koan commented on lib/on_form/form.rb:
```diff
@@ -91,8 +91,13 @@ def self.expose_collection_of(association_name, on: nil, prefix: nil, suffix: ni
+      end
+
       define_method("#{exposed_name}_attributes=") { |params| send(exposed_name).parse_collection_attributes(params) }
+      define_method("_validate_#{exposed_name}_forms") { send(exposed_name).validate_forms(self) }
```
If we go with this approach, I'd rather do something like

```
send(exposed_name).validation_errors.each { |error| self.errors.add(exposed_name, error) }
```

rather than add more coupling between self and the collection wrapper

Dhamsoft replied:
Yep, we can do that. Another problem here adding child form validation and child form backing model validation adding here. Before adding parent model backing validation error. Its looking odd while displaying the error message.

I'm trying to solve this problem on https://github.com/fluxfederation/on_form/pull/6. 
ie: Always run all form validation first then go for all backing model validation.
